lib/engine/step/g_1860/buy_cert.rb

```diff
@@ -0,0 +1,268 @@
+          setup_auction
+        end
+
+        def process_pass(action)
+          player = action.entity
+
+          @log << "#{player.name} passes bidding"
+
+          @bids.delete(player)
+
+          resolve_auction
+        end
+
+        def process_bid(action)
+          player = action.entity
+          price = action.price
+
+          if !@in_auction
+            buy_company(player, action.company, price)
+          else
+            if price > max_player_bid(player)
+              @game.game_error("Cannot afford bid. Maximum possible bid is #{max_player_bid(player)}")
+            end
+
+            @log << "#{player.name} bids #{@game.format_currency(price)}"
+
+            @bids[player] = price
+          end
+        end
+
+        def get_par_prices(entity, corp)
+          prices = @game.par_prices(corp).select { |p| p.price * 2 <= entity.cash }
+          if prices.empty? && cheapest_thing.corporation?
+            # assumes all corps available have same minimum par price
+            return [@game.par_prices(corp).min_by(&:price)]
+          end
+
+          prices
+        end
+
+        def active_entities
+          return [@bids.min_by { |_k, v| v }.first] if @in_auction
+
+          super
+        end
+
+        def min_increment
+          1
+        end
+
+        def min_player_bid
+          highest_player_bid + MIN_BID_RAISE
+        end
+
+        def max_player_bid(entity)
+          entity.cash - cheapest_price
+        end
+
+        def min_bid(company)
+          return unless company
+
+          company.value
+        end
+
+        def companies_pending_par
+          false
+        end
+
+        def visible
+          true
+        end
+
+        def committed_cash(player, _show_hidden = false)
+          if @bids[player] && !@bids[player].negative?
+            @bids[player] + cheapest_price
+          else
+            0
+          end
+        end
+
+        private
+
+        def highest_player_bid
+          @in_auction && any_bids? ? @bids.max_by { |_k, v| v }.last : 0
+        end
+
+        def highest_bid
+          @in_auction ? @bids.max_by { |_k, v| v }.last : 0
+        end
+
+        def any_bids?
+          @in_auction && @bids.max_by { |_k, v| v }.last.positive?
+        end
+
+        def cheapest_thing
+          @companies.min_by { |c| c.company? ? c.value : @game.par_prices(c).map(&:price).min * 2 }
+        end
+
+        def cheapest_price
+          thing = cheapest_thing
+          thing.company? ? thing.value : @game.par_prices(thing).map(&:price).min * 2
+        end
+
+        def setup_auction
+          @bids.clear
+          @first_player = current_entity
+          start_idx = entity_index
+          size = entities.size
+          # initialize bids to preserve player order starting with current player
+          entities.each_index do |idx|
+            @bids[entities[idx]] = -size + (idx - start_idx) % size
+          end
+          @in_auction = true
```

Comment:
shouldn't you just have one method that is like auctioning?

Reply:
I don't understand. Are you suggesting I don't need setup_auction ? Or that I shouldn't be checking @in_action directly?

Reply:
is there a way where you can determine in_auction without having to set it?

Reply:
if not, just call it @auctioning and make it attr_reader

Reply:
Ok